api/src/main/java/javax/enterprise/inject/spi/builder/BeanConfigurator.java

```diff
@@ -0,0 +1,355 @@
+
+    /**
+     *
+     * Add InjectionPoints to the configured bean
+     *
+     * @param injectionPoints the injectionPoints to add
+     * @return self
+     */
+    BeanConfigurator<T> addInjectionPoints(Set<InjectionPoint> injectionPoints);
+
+    /**
+     *
+     * Replace InjectionPoints for the configured bean
+     *
+     * @param injectionPoints the injectionPoints for the configured bean
+     * @return self
+     */
+    BeanConfigurator<T> injectionPoints(InjectionPoint... injectionPoints);
+
+    /**
+     *
+     * Replace InjectionPoints for the configured bean
+     *
+     * @param injectionPoints the injectionPoints for the configured bean
+     * @return self
+     */
+    BeanConfigurator<T> injectionPoints(Set<InjectionPoint> injectionPoints);
+
+    /**
+     *
+     * Make the configured bean implements {@link PassivationCapable} and its Id for passivation.
+     *
+     *
+     * @param id for
+     * @see PassivationCapable#getId()
+     * @return self
+     */
+    BeanConfigurator<T> id(String id);
+
+    /**
+     *
+     * Set a {@link Function} to create a bean instance from a {@link CreationalContext}
+     *
+     * @param callback the Function to create the instance
+     * @return self
+     */
+    <U extends T> BeanConfigurator<U> createWith(Function<CreationalContext<U>, U> callback);
+
+    /**
+     *
+     * Set a {@link Supplier} to create a bean instance
+     *
+     *
+     * @param callback the Supplier to create the instance
+     * @return self
+     */
+    <U extends T> BeanConfigurator<U> produceWith(Supplier<U> callback);
+
+    /**
+     * Set a {@link Function} to create a bean instance from an {@link Instance}
+     *
+     * @param callback the function to create the instance
+     * @return self
+     */
+    <U extends T> BeanConfigurator<U> produceWith(Function<Instance<Object>, U> callback);
+
+    /**
+     * A shortcut for {@code produceWith(() -> existing} where {@code existing} represents an instance whose lifecycle is not managed by CDI.
+     *
+     * @param instance use as produced instance for the configured bean
+     * @return self
+     */
+    <U extends T> BeanConfigurator<U> producing(U instance);
+
+    /**
+     *
+     * Set a {@link BiConsumer} to destroy a bean instance from a {@link CreationalContext}.
+     * If no destroy callback is specified, a NOOP dispose callback is automatically set.
+     *
+     * @param callback the BiConsumer to destroy the instance
+     * @return self
+     */
+    BeanConfigurator<T> destroyWith(BiConsumer<T, CreationalContext<T>> callback);
+
+    /**
+     *
+     * Set a {@link Consumer} to destroy a bean instance.
+     * If no dispose callback is specified, a NOOP dispose callback is automatically set.
+     *
+     * @param callback the Consumer to dispose the instance
+     * @return self
+     */
+    BeanConfigurator<T> disposeWith(Consumer<T> callback);
```

Comment:
Shouldn't ```disposeWith``` and ```destroyWith``` also return ```<U>``` rather than ```<T>```?
E.g. ```<U extends T> BeanConfigurator<U> disposeWith(Consumer<U> callback);```
I am rather asking than saying this is wrong. It just caught my attention :)